datasets/ua/war_sanctions/crawler.py

```diff
@@ -0,0 +1,348 @@
+
+
+LINKS = [
+    {  # child kidnappers
+        "url": "https://war-sanctions.gur.gov.ua/en/kidnappers/persons?page=1&per-page=12",
+        "type": "person",
+    },
+    {  # child kidnappers
+        "url": "https://war-sanctions.gur.gov.ua/en/kidnappers/companies?page=1&per-page=12",
+        "type": "company",
+    },
+    {  # russian athletes
+        "url": "https://war-sanctions.gur.gov.ua/en/sport/persons?page=1&per-page=12",
+        "type": "person",
+    },
+    {  # ships
+        "url": "https://war-sanctions.gur.gov.ua/en/transport/ships?page=1&per-page=12",
+        "type": "vessel",
+    },
+]
+
+# TODO: c/o and 'Unknown (22.03.2024), Ship Safety Management Manager (IMO / Country / Date)': 'Unknown (08.02.2023)
+# TODO: Sanction Jurisdictions
+
+
+def lookup_override(context, key):
+    override_res = context.lookup("overrides", key)
+    if not override_res:
+        context.log.warning(f"No override found for {key}")
+        return {}
+
+    extracted = {}
+    for override in override_res.items:
+        if override["prop"] == "name":
+            extracted["name"] = override["value"]
+        elif override["prop"] == "registrationCode":
+            extracted["registrationCode"] = override["value"]
+
+    if not extracted:
+        context.log.warning(f"No matching properties found in overrides for {key}")
+
+    return extracted
+
+
+def extract_label_value_pair(label_elem, value_elem, data):
+    label = label_elem.text_content().strip().replace("\n", " ")
+    value = [text.strip() for text in value_elem.itertext() if text.strip()]
+    if len(value) == 1:
+        value = value[0]
+    data[label] = value
+    return label, value
+
+
+def crawl_index_page(context: Context, index_page, data_type):
+    index_page = context.fetch_html(index_page, cache_days=3)
+    main_grid = index_page.find('.//div[@id="main-grid"]')
+    if main_grid is not None:
+        for link in main_grid.xpath(".//a/@href"):
+            if link.startswith("https:"):
+                if data_type == "person":
+                    crawl_person(context, link)
+                elif data_type == "company":
+                    crawl_company(context, link)
+                if data_type == "vessel":
+                    crawl_vessel(context, link)
+
+
+def crawl_vessel(context: Context, link):
+    detail_page = context.fetch_html(link, cache_days=3)
+    details_container = detail_page.find(".//main")
+    data: dict[str, str] = {}
+
+    xpath_definitions = [
+        (
+            "main_info_rows",
+            ".//div[contains(@class,'tools-spec')]/div[contains(@class, 'row')]",
+        ),
+        (
+            "justification_rows",
+            ".//div[contains(@class, 'tools-frame')]/div[contains(@class, 'mb-3')]",
+        ),
+        (
+            "additional_info_rows",
+            ".//div[contains(@class, 'tools-frame')]//div[@class='mb-3' or contains(@class, 'js_visibility')]",
+        ),
+    ]
```

Comment:
I don't really see a better way to do this. These are potentially brittle classes that could be used anywhere else too, but this is scoped to tools-frame and the .pop()s below should alert us to changes that make these lookups invalid.

👍